code/gateway/deregister.q

```diff
@@ -0,0 +1,52 @@
+
+\d .gw
+
+deregcheckfreq:@[value;`.gw.deregcheckfreq;0D00:00:10];
+
+//overwrite this method to not upsert null handles??
+addserversfromconnectiontable:{
+ {.gw.addserverattr'[x`w;x`proctype;x`attributes]}[select w,proctype,attributes from .servers.SERVERS where ((proctype in x) or x~`ALL),not w in ((0;0Ni),exec handle from .gw.servers where not null handle)];}
+
+\d .finspace
+
+unregisterfromgw:{[servernames]
+   // identify the serverid by mapping the handles in .server.SERVERS and .gw.servers
+   svrIDs:exec serverid from deregservers:(select handle:w, procname from .servers.SERVERS where procname in servernames, not null w) ij `handle xkey .gw.servers;
+   if[not count svrIDs; .lg.o[`unregisterfromgw;"No servers to deregister. Aborting"]; :()];
+
+   // block incoming queries to these processes
+   update active:0b,disconnecttime:.proc.cp[] from `.gw.servers where serverid in svrIDs;
+   
+   timerID:first 1?0Ng;
+   .timer.repeat[.proc.cp[];0Wp;.gw.deregcheckfreq;(`.finspace.checkremainingqueries;timerID;svrIDs;update flagdown:0b from deregservers);"check if pending queries in servernames"]
+ };
+
+// timerid     | -2h | guid identifier to help remove instance of this function if multiple instances of this function are scheduled
+// serversid   | 6h  | ids of the servers to deregister
+// serversinfo | 98h | table with info about the servers to deregister
+checkremainingqueries:{[timerid;serverids;serversinfo]
+   whereComplete:serverids where checkremainingqueriesforserver each serverids;
+   serversinfo:update flagdown:1b from serversinfo where serverid in raze @'[.finspace.sigserverexit[;serversinfo];whereComplete;{:()}];
+
+   if[all serversinfo`flagdown;
+      .timer.remove @/: exec id from .timer.timer where `.finspace.checkremainingqueries in' funcparam, timerid in' funcparam;
+      :(::);
+     ]; 
+
+   if[count incompleteservers:exec procname from serversinfo where not serverid in whereComplete;
+      .lg.o[`checkremainingqueries;"servers ",(.Q.s1 incompleteservers), " still have remaining queries"]];
+ };
```

Comment:
it's not clear to me what this function actually does - I can see it removes itself if all queries are complete, and it logs a message if they aren't, but is any of that actually necessary? is the purpose here purely to log if there are still queries in-flight? if so, do we care?

Reply:
this function will signal a shutdown for a server where all queries are complete. It will then mark that particular server as "down". It will do this until all servers identified by the parameter ```serverids``` are marked as "down", and then remove itself from the timer

Reply:
Code has been updated to use the callback approach instead of the timer based approach